wire-ios-data-model/Proteus/ProteusService.swift

```diff
@@ -27,11 +27,297 @@ public final class ProteusService: ProteusServiceInterface {
+            )
+            return encryptedBatch.mapValues(\Bytes.data)
+        } catch {
+            logger.error("failed to encrypt data batch: \(String(describing: error))")
+            throw EncryptionError.failedToEncryptDataBatch
+        }
+    }
+
+    // MARK: - proteusDecrypt
+
+    enum DecryptionError: Error {
+        case failedToDecryptData
+    }
+
+    public func decrypt(data: Data, forSession id: String) throws -> Data {
+        logger.info("decrypting data")
+
+        do {
+            let decryptedBytes = try coreCrypto.proteusDecrypt(
+                sessionId: id,
+                ciphertext: data.bytes
+            )
+            return decryptedBytes.data
+        } catch {
+            logger.error("failed to decrypt data: \(String(describing: error))")
+            throw DecryptionError.failedToDecryptData
+        }
+    }
+
+    // MARK: - proteusNewPrekey
+
+    enum PrekeyError: Error {
+        case failedToGeneratePrekey
+        case prekeyCountTooLow
+        case prekeyCountTooHigh
+        case failedToGetLastPrekey
+    }
+
+    public func generatePrekey(id: UInt16) throws -> String {
+        logger.info("generating prekey")
+
+        do {
+            return try coreCrypto.proteusNewPrekey(prekeyId: id).base64EncodedString
+        } catch {
+            logger.error("failed to generate prekey: \(String(describing: error))")
+            throw PrekeyError.failedToGeneratePrekey
+        }
+    }
+
+    public func lastPrekey() throws -> String {
+        logger.info("getting last resort prekey")
+        do {
+            return try coreCrypto.proteusLastResortPrekey().base64EncodedString
+        } catch {
+            logger.error("failed to get last resort prekey: \(String(describing: error))")
+            throw PrekeyError.failedToGetLastPrekey
+        }
+    }
+
+    public var lastPrekeyID: UInt16 {
+        let lastPrekeyID = try? coreCrypto.proteusLastResortPrekeyId()
+        return lastPrekeyID ?? UInt16.max
+    }
+
+    public func generatePrekeys(count: UInt16 = 0, start: UInt16 = 0) throws -> [IdPrekeyTuple] {
+        guard count > 0 else {
+            throw PrekeyError.prekeyCountTooLow
+        }
+
+        guard start + count < lastPrekeyID else {
+            throw PrekeyError.prekeyCountTooHigh
+        }
+
+        let range = prekeysRange(count, start: start)
+        let prekeys = try generatePrekeys(range)
+
+        guard prekeys.count > 0 else {
+            throw PrekeyError.failedToGeneratePrekey
+        }
+
+        return prekeys
+    }
+
+    private func generatePrekeys(_ range: CountableRange<UInt16>) throws -> [IdPrekeyTuple] {
+        return try range.map {
+            let prekey = try generatePrekey(id: $0)
+            return (id: $0, prekey: prekey)
+        }
+    }
+
+    private func prekeysRange(_ count: UInt16, start: UInt16) -> CountableRange<UInt16> {
+        if start > lastPrekeyID - count {
+            return 0 ..< count
+        }
```

Comment:
Could you explain this logic? I don't quite understand it means.

Reply:
https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/680951853/Use+case+uploading+Proteus+prekeys
> The safest way to generate prekeys ID is to always generate increasing IDs, until we reach the limit of 65535. This will take a while, so much that once we reach that limit, we can almost be sure that whoever has consumed prekeys 0, 1, 2… had enough time to send us their prekey messages (or will never do it). Therefore we can restart from 0 after reaching 65535.

I replicated [this logic](https://github.com/wireapp/wire-ios-mono/blob/5ad304cb247d200994e62c8433d15af807f6a1e4/wire-ios-data-model/Source/Utilis/CryptoBox.swift#L153) from `UserClientKeyStore`. My understanding is that if we reach the max limit of prekeys, we start over from zero.